### ISSUE TITLE
SPIKE: Evaluate TF Plans for high risk changes and update review processes

### ISSUE THREAD

**Issue by richgreen-moj** [ministryofjustice/modernisation-platform#6388]
### User Story

As a MP Engineer
I want to add more unit testing for TF Plans using Conftest or similar technology
So that I can either block changes, warn PR reviewers, or request additional reviewers before changes are merged to main

### Value / Purpose

Building upon the work in https://github.com/ministryofjustice/modernisation-platform/pull/6281 , this issue would take it further to start evaluating the output of Terraform plans and writing policies/tests that would either prevent or warn reviewers of high risk changes to _specific resources_.

- Deleting certain threshold of TF resources
- Deleting a specific type of resource
- Updating a major version of a module

The result of these tests could block a change, add more detailed warnings for reviewers or even request another level of review.

It could also look at low risk changes and make it quicker for these kinds of changes to be merged. 

One idea could be that we default to changes requiring 2 reviewers but if the changes are deemed low risk then the CI pipeline reduces required reviews to 1.

### Useful Contacts

@richgreen-moj

### Definition of Done

- [x] Agree timescale for SPIKE - limit to 3 days
- [x] Experiment with writing policies that look for high risk changes to TF resources
- [x] Work with team to consider whether a new branch protection rule would be appropriate for high risk change
- [x] User docs have been updated 
- [x] Another team member has reviewed

**Comment by richgreen-moj** [comment]
Some options to consider:

1 - Create a script that is added to all workflows running terraform which evaluates the plan and adds comments to the PR with warnings if sensitive resources are due to be deleted or a high number of destructive changes are due to be made (e.g. over X amount of resources will be deleted). The possibilities of what to warn for are endless and could be added to over time.  This would be easy to implement and low impact on workflow (i.e. it's just a warning).

2 - Simulate the [TF Plan Evaluator script](https://github.com/ministryofjustice/modernisation-platform-environments/blob/main/scripts/terraform-plan-evaluator.sh) and [logic](https://github.com/ministryofjustice/modernisation-platform-environments/blob/main/.github/workflows/reusable_terraform_plan_apply.yml#L205) used in the MP Environments repo which won't allow you to apply TF changes that are highlighted in the script unless a second GH team has reviewed the PR (i.e. MP team). This works on the basis that the plan and apply steps are linked in a workflow `reusable_terraform_plan_apply.yml` and the job 'fails' at the plan stage until a specified extra reviewer group has approved the PR. Once the extra review is done the 'apply' phase will run on a subsequent re-running of the workflow.

To do this on MP we would need to create a new GH team (maybe called "MP Reviewers") with either all the existing MP members (or a select few if desired e.g. architects/leads?) which would need to approve certain PRs if the changes were deemed risky by the evaluator script. Essentially it would up the required reviewers to 2. 
(**In theory** you could have a script that would temporarily change the number of required reviewers in the repository settings but as we have so many workflows that can be running concurrently I don't think it is a suitable strategy).

There's not really an easy way currently to dynamically change the amount of reviewers required to merge a PR. You can use a CODEOWNERS file and reference multiple reviewers but that's still tied to the overall number of required reviewers (currently set to 1 in our repo)

**Comment by richgreen-moj** [comment]
I ran a session with the team today to show what I had done for the spike. I demoed the OPA tests I'd written that query the json output of TF plans and error if;

- a threshold of resources are marked for delation (e.g. > 50)
- a list of sepcidied resources are marked for deletion or
- resources that match a particular tag are marked for deletion

We then discussed the two options listed in the previous comment;

1 - People seemed to think that warning messages in PRs would be good and also potentially the ability to output a TF plan into the PR message would be useful as it saves you having to dive into logs.

2 - I described the potential for us copying the process in MP Envs where certain changes are evaluated and require an extra review and whether we should do similar. This was not popular as it could be hard to decide on what kinds of changes require extra reviews, potentially all the changes we make can have quite an impact so it might just be too disruptive to the workflow. 
I also briefly explained a bug which I think means that currently members can merge in changes we want MP to review as the workflow only warns and raises a comment it doesn't change the threshold of required reviews in branch protection rules. We agreed that it should be fixed especially if it is as simple as adding a required status check.
We also agreed that a reusable TF plan/apply job like the one in MP Envs repo might be worth applying to our MP workflows just so it is easier to maintain and build in extra steps e.g. the ability to post plans to the PR or other warnings.

Outcomes:
- Write story for creating a re-usable terraform plan/apply job for use across the various MP Terraform workflows
   - it should have ability to post plan output to PRs
   - it could have an extra warning message on PR if certain resources are being deleted
- Raise a bug to fix potential issue in MP Envs repo where members can merge PRs to main without MP approval
